icicle-compiler/src/Icicle/Sea/FromAvalanche/Base.hs

```diff
@@ -35,25 +40,75 @@ import           Text.Printf (printf)
 
 ------------------------------------------------------------------------
 
-textOfName :: Pretty n => n -> Text
-textOfName = T.pack . show . seaOfName
-
-seaOfName :: Pretty n => n -> Doc
-seaOfName = string . fmap mangle . show . pretty
+-- | A legal C identifier.
+newtype SeaName = SeaName {
+    getSeaName :: Text
+ } deriving (Eq, Ord, Show)
+
+-- | A value that will need be quoted in C.
+newtype SeaString = SeaString {
+    getSeaString :: Text
+ } deriving (Eq, Ord, Show)
+
+instance Pretty SeaName where
+  pretty =
+    string . Text.unpack . takeSeaName
+
+instance Pretty SeaString where
+  pretty =
+    string . Text.unpack . getSeaString
+
+takeSeaName :: SeaName -> Text
+takeSeaName = getSeaName
+
+seaNameValidHead :: Char -> Bool
+seaNameValidHead c =
+  c >= 'a' && c <= 'z'
+{-# INLINE seaNameValidHead #-}
+
+seaNameValidTail :: Char -> Bool
+seaNameValidTail c =
+  seaNameValidHead c ||
+  (c >= '0' && c <= '9')
+{-# INLINE seaNameValidTail #-}
+
+asSeaName :: Text -> Maybe SeaName
+asSeaName t =
+  case Text.unpack t of
+    x:xs | seaNameValidHead x && all seaNameValidTail xs ->
+      Just (SeaName t)
+    _ ->
+      Nothing
+
+mangleToSeaName :: Pretty n => n -> SeaName
+mangleToSeaName = SeaName . Text.pack . concatMap mangle . show . pretty
   where
-    mangle '$' = '_'
-    mangle ' ' = '_'
-    mangle  c  =  c
-
-seaOfNameIx :: Pretty n => n -> Int -> Doc
-seaOfNameIx n ix = seaOfName (pretty n <> text "$ix$" <> int ix)
-
-seaOfAttributeDesc :: Attribute -> Doc
-seaOfAttributeDesc (Attribute xs)
-  | T.null xs = string ""
-  | otherwise = pretty (T.filter isLegal xs)
- where
-  isLegal c = isLower c || isUpper c || c == ' ' || c == '_'
+    mangle c
+     | c >= '0' && c <= '9'
+     = [c]
+     | c >= 'a' && c <= 'z'
+     = [c]
+     | c >= 'A' && c <= 'Z'
+     = [c]
+     | c == '_'
+     = "_"
+     | otherwise
+     = "_" <> showHex (ord c) ""
```

Comment:
I think this function needs a test, I suspect it is not generating unambiguous names.